# Association Rule Mining in Mass-Spectrometry Data
### *Muhammad Haseeb and Muhammad Usman Tariq*
##### *CAP6778: Adv. Topics in Data Mining - Fall 2019*
##### *School of Computing and Information Sciences (SCIS)*
##### *Florida International University (FIU)*

### The MIT License

##### *Copyright 2019 Muhammad Haseeb and Muhammad Usman Tariq*

Permission is hereby granted, free of charge, to any person obtaining a copy of this software 
and associated documentation files (the "Software"), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, 
and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial 
portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT 
LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## Required Imports

In [1]:
# Import the required libraries
import os
import re
import sys
import math
import time
import urllib
import operator
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from IPython.display import clear_output

## Data Acquisition
Let's get the annotated HUMAN MS/MS spectral data libraries from <a href="https://chemdata.nist.gov/dokuwiki/doku.php?id=peptidew:cdownload" class="external">NIST.</a> Note that this can be done for any spectral library or multiple spectral libraries at once as well.

In [2]:
mouse = 'cptac2_mouse_hcd_selected.msp.tar.gz'
msp_file = './data/cptac2_mouse_hcd_selected.msp'
data_dir = os.path.expanduser("./data")

if not os.path.isdir(data_dir):
    os.mkdir(data_dir)

if not os.path.exists(msp_file):
    if not os.path.exists(mouse):
        !wget 'ftp://chemdata.nist.gov/download/peptide_library/libraries/cptaclib/2015/cptac2_mouse_hcd_selected.msp.tar.gz'
    
    # Extract the tar ball into MSP file and move into the speclib folder
    !tar xvf 'cptac2_mouse_hcd_selected.msp.tar.gz'
    os.rename('./cptac2_mouse_hcd_selected.msp', msp_file)

### Data Extraction Hyperparameters

In [3]:
# Data extraction Hyperparamters
divider = 5                     # Divide every m/z by this number.
max_len = int(5000 / divider)  # Max allowed m/z value.
filter_peaks = 100              # How many peaks to keep after filtration.

In [4]:
'''To keep updating this dict for multiple files. Dont define it again.'''
pep_spec = {}

### MSP File Parser

Read the MSP file, extract the required ions and pre-prcocess spectra

In [5]:
f=open(msp_file, "r")
lines = f.readlines()
f.close()

isName = isMW = isNumPeaks = False
new = prev = 0
i = 0


while i < len(lines):
    '''
    Process each line of the file.
    '''
    line = lines[i]
    i += 1
    
    splits = line.split(':')
    
    '''Keep going through lines. Name line is peptide and
    all the lines after Num peaks are the spectrum'''
    if splits[0] == 'Name':
        split1 = splits[1]
        pep = split1.split('/')[0].lstrip(' ')
        isName = True
        
    if isName and splits[0] == 'MW':
        mass = float(splits[1]) # Not using it anywhere. It's just there.
        isMW = True
        
    if isName and isMW and splits[0] == 'Num peaks':
        '''After this line, we will start processing the spectrum.'''
        numPeaks = int(splits[1]) # Just in case.
        
        temp_spec = np.zeros(max_len)
        while (lines[i] != '\n'):
            mzline = lines[i]
            i += 1
            mzsplits = mzline.split('\t')
            # Get the m/z and intenisty value of each line.
            moz, intensity, pattern = float(mzsplits[0]), float(mzsplits[1]), mzsplits[2]

            # Filter out only the non-noisy terminal ions from the spectrum
            lbls = re.findall(r"(?<!\/)(?<![A-Z])([abcxyz][0-9]+)", pattern)

            if lbls:
                # Dividing m/z by the divider.
                temp_spec[round(moz / divider)] = intensity

        isNumPeaks = True
        
    if isName and isMW and isNumPeaks:
        '''At this point, we are done reading one spectrum.
        Place everything where it needs to go and read the next one.'''
        isName = isMW = isNumPeaks = False
        
        # This is the filter. Gets the top 100 peaks
        # and returns them in one hot encoded vector of size max_len
        top_indx = np.argpartition(temp_spec, -filter_peaks)[-filter_peaks:]
        spec = np.zeros(max_len, dtype=bool)
        spec[top_indx] = True
        
        # Creating dictionary. Key: peptide, Value: spectrum.
        # As a side effect we'll get rid of duplicates.
        if pep not in pep_spec: pep_spec[pep] = spec
        
        # Displays the progress.
        new = int((i/len(lines)) * 100)
        if new > prev:
            clear_output(wait=True)
            print(str(new) + '%')
            prev = new


99%


## Frequent Itemset Mining
Construct the market-basket dataframe.

### Frequent Itemset Hyperparameters

In [6]:
max_itemset_size = 5     # Max size of the frequent itemsets.
min_support = 0.40       # Maybe we should increase this?

### DataFrame Construction

In [7]:
# Just get the value from the dict.
buckets = list(pep_spec.values())
print(len(buckets))

# The apriori function needs a dataframe.
df_buckets = pd.DataFrame(buckets, columns=list(range(0, max_len)))

10026


In [8]:
df_buckets

0      1      2      3      4      5      6      7      8      9    \
0      False  False  False  False  False  False  False  False  False  False   
1      False  False  False  False  False  False  False  False  False  False   
2      False  False  False  False  False  False  False  False  False  False   
3      False  False  False  False  False  False  False  False  False  False   
4      False  False  False  False  False  False  False  False  False  False   
5      False  False  False  False  False  False  False  False  False  False   
6      False  False  False  False  False  False  False  False  False  False   
7      False  False  False  False  False  False  False  False  False  False   
8      False  False  False  False  False  False  False  False  False  False   
9      False  False  False  False  False  False  False  False  False  False   
10     False  False  False  False  False  False  False  False  False  False   
11     False  False  False  False  False  False  False  False  False  False   
12     False  False  False  False  False  False  False  False  False  False   
13     False  False  False  False  False  False  False  False  False  False   
14     False  False  False  False  False  False  False  False  False  False   
15     False  False  False  False  False  False  False  False  False  False   
16     False  False  False  False  False  False  False  False  False  False   
17     False  False  False  False  False  False  False  False  False  False   
18     False  False  False  False  False  False  False  False  False  False   
19     False  False  False  False  False  False  False  False  False  False   
20     False  False  False  False  False  False  False  False  False  False   
21     False  False  False  False  False  False  False  False  False  False   
22     False  False  False  False  False  False  False  False  False  False   
23     False  False  False  False  False  False  False  False  False  False   
24     False  False  False  False  False  False  False  False  False  False   
25     False  False  False  False  False  False  False  False  False  False   
26     False  False  False  False  False  False  False  False  False  False   
27     False  False  False  False  False  False  False  False  False  False   
28     False  False  False  False  False  False  False  False  False  False   
29     False  False  False  False  False  False  False  False  False  False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
9996   False  False  False  False  False  False  False  False  False  False   
9997   False  False  False  False  False  False  False  False  False  False   
9998   False  False  False  False  False  False  False  False  False  False   
9999   False  False  False  False  False  False  False  False  False  False   
10000  False  False  False  False  False  False  False  False  False  False   
10001  False  False  False  False  False  False  False  False  False  False   
10002  False  False  False  False  False  False  False  False  False  False   
10003  False  False  False  False  False  False  False  False  False  False   
10004  False  False  False  False  False  False  False  False  False  False   
10005  False  False  False  False  False  False  False  False  False  False   
10006  False  False  False  False  False  False  False  False  False  False   
10007  False  False  False  False  False  False  False  False  False  False   
10008  False  False  False  False  False  False  False  False  False  False   
10009  False  False  False  False  False  False  False  False  False  False   
10010  False  False  False  False  False  False  False  False  False  False   
10011  False  False  False  False  False  False  False  False  False  False   
10012  False  False  False  False  False  False  False  False  False  False   
10013  False  False  False  False  False  False  False  False  False  False   
10014  False  False  False  False  False  False  False  Fal

### Apriori Algorithm
Using the association rule mining algorithms from: *Agrawal, Rakesh, and Ramakrishnan Srikant. "Fast algorithms for mining association rules." Proc. 20th int. conf. very large data bases, VLDB. Vol. 1215. 1994*. Using implementation from the [mlxtend.apriori](http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/) Python package.

In [9]:
freq_itemsets = apriori(df_buckets,
                        min_support=min_support, 
                        max_len=max_itemset_size, 
                        verbose=1,
                        low_memory=True)

Iteration: 7083542 | Sampling itemset size 5


### Freq Itemsets Visualization

Now that we have mined the frequent itemsets from the dataframe, let's visualize some stats and see if we need to refine any hyperparameters and mine itemsets again.

In [10]:
# Add another column to store lengths of the frequent itemsets
freq_itemsets['length'] = freq_itemsets['itemsets'].apply(lambda x: len(x))

# Filter out the small itemsets (of length < 3)
filtered_freqITS = freq_itemsets[ (freq_itemsets['length'] > 2) &
                                        (freq_itemsets['support'] > 0.8) ]
# We found 625 filtered frequent itemsets
disp_ffITS = filtered_freqITS.reset_index().rename(columns={'index' : 'itemset#'})

filtered_freqITS = filtered_freqITS.drop(['length'], axis=1)

In [11]:
disp_ffITS

itemset#   support                   itemsets  length
0       36387  0.804608            (352, 353, 351)       3
1       36388  0.803212            (352, 354, 351)       3
2       36412  0.803511            (353, 354, 351)       3
3       36413  0.800020            (353, 355, 351)       3
4       36436  0.800319            (354, 355, 351)       3
5       36711  0.809595            (352, 353, 354)       3
6       36712  0.805506            (352, 353, 355)       3
7       36713  0.803012            (352, 353, 356)       3
8       36714  0.801217            (352, 353, 357)       3
9       36715  0.800818            (352, 353, 358)       3
10      36735  0.806802            (352, 354, 355)       3
11      36736  0.803910            (352, 354, 356)       3
12      36737  0.802115            (352, 354, 357)       3
13      36738  0.801815            (352, 354, 358)       3
14      36739  0.800319            (352, 354, 359)       3
15      36758  0.803810            (352, 355, 356)       3
16      36759  0.801217            (352, 355, 357)       3
17      36760  0.800120            (352, 355, 358)       3
18      36780  0.800519            (352, 356, 357)       3
19      36821  0.800020            (352, 358, 359)       3
20      37010  0.814383            (353, 354, 355)       3
21      37011  0.811191            (353, 354, 356)       3
22      37012  0.809894            (353, 354, 357)       3
23      37013  0.807600            (353, 354, 358)       3
24      37014  0.806503            (353, 354, 359)       3
25      37015  0.805705            (360, 353, 354)       3
26      37016  0.804010            (353, 354, 361)       3
27      37017  0.802713            (353, 354, 362)       3
28      37018  0.802314            (353, 354, 363)       3
29      37019  0.801416            (353, 354, 364)       3
..        ...       ...                        ...     ...
726   5201533  0.802713  (358, 360, 361, 363, 364)       5
727   5201534  0.800020  (358, 360, 361, 363, 365)       5
728   5201637  0.802513  (358, 360, 362, 363, 364)       5
729   5202182  0.802912  (358, 361, 362, 363, 364)       5
730   5204444  0.809894  (359, 360, 361, 362, 363)       5
731   5204445  0.806703  (359, 360, 361, 362, 364)       5
732   5204446  0.802414  (359, 360, 361, 362, 365)       5
733   5204459  0.805306  (359, 360, 361, 363, 364)       5
734   5204460  0.802115  (359, 360, 361, 363, 365)       5
735   5204473  0.800219  (359, 360, 361, 364, 365)       5
736   5204563  0.804907  (359, 360, 362, 363, 364)       5
737   5204564  0.801017  (359, 360, 362, 363, 365)       5
738   5204667  0.800120  (359, 360, 363, 364, 365)       5
739   5205108  0.805306  (359, 361, 362, 363, 364)       5
740   5205109  0.801716  (359, 361, 362, 363, 365)       5
741   5205212  0.800120  (359, 361, 363, 364, 365)       5
742   5205653  0.800219  (359, 362, 363, 364, 365)       5
743   5207370  0.809795  (360, 361, 362, 363, 364)       5
744   5207371  0.806004  (360, 361, 362, 363, 365)       5
745   5207384  0.804010  (360, 361, 362, 364, 365)       5
746   5207474  0.804109  (360, 361, 363, 364, 365)       5
747   5207915  0.804708  (360, 362, 363, 364, 365)       5
748   5209632  0.808398  (361, 362, 363, 364, 365)       5
749   5209633  0.803112  (361, 362, 363, 364, 366)       5
750   5209645  0.801815  (361, 362, 363, 365, 366)       5
751   5209722  0.802115  (361, 362, 364, 365, 366)       5
752   5210073  0.803112  (361, 363, 364, 365, 366)       5
753   5211349  0.806104  (362, 363, 364, 365, 366)       5
754   5211350  0.800618  (362, 363, 364, 365, 367)       5
755   5212625  0.802912  (363, 364, 365, 366, 367)       5

[756 rows x 4 columns]

We found about 756 frequent itemsets with support > 0.8 and length > 2. These look good. Let's mine the association rules.

## Association Rule Mining

Let's mine association rules from the obtained frequent itemsets and take a look at rules ($R = I\rightarrow j$) with high *confidence* ($Conf$) and *interest* ($Interest$):

$$
Conf(I\rightarrow j) = \frac{Support(I \cup j)}{Support(I)} \\
Interest(I \rightarrow j) = Conf(I\rightarrow j) - Pr [j]
$$
where $Pr [j]$ is the probability of the item $j$ computed as: $support(j)$

### Association Rules Hyperparameters

In [24]:
confidence = 0.80     # Association Rule Confidence
intr   = 0.50         # Association Rule Interest (leverage)

### Association Rules
Using the association rule mining algorithms from: *Agrawal, Rakesh, and Ramakrishnan Srikant. "Fast algorithms for mining association rules." Proc. 20th int. conf. very large data bases, VLDB. Vol. 1215. 1994*. Using implementation from the [mlxtend.association_rules](http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/) Python package.

In [13]:
# Get association rules based on the confidence 
rules_conf = association_rules(freq_itemsets, 
                               metric='confidence', 
                               min_threshold=confidence, 
                               support_only=False)

In [14]:
# Get association rules based on the interest
rules_conf['interest'] = rules_conf['confidence'] - rules_conf['consequent support']

### Association Rules Visualization

In [15]:
rules_conf

antecedents      consequents  antecedent support  \
0                         (26)             (29)            0.486036   
1                         (26)            (345)            0.486036   
2                         (26)            (346)            0.486036   
3                         (26)            (347)            0.486036   
4                         (26)            (348)            0.486036   
5                         (26)            (349)            0.486036   
6                         (26)            (350)            0.486036   
7                         (26)            (351)            0.486036   
8                         (26)            (352)            0.486036   
9                         (26)            (353)            0.486036   
10                        (26)            (354)            0.486036   
11                        (26)            (355)            0.486036   
12                        (26)            (356)            0.486036   
13                        (26)            (357)            0.486036   
14                        (26)            (358)            0.486036   
15                        (26)            (359)            0.486036   
16                        (26)            (360)            0.486036   
17                        (26)            (361)            0.486036   
18                        (26)            (362)            0.486036   
19                        (26)            (363)            0.486036   
20                        (26)            (364)            0.486036   
21                        (26)            (365)            0.486036   
22                        (26)            (366)            0.486036   
23                        (26)            (367)            0.486036   
24                        (26)            (368)            0.486036   
25                        (26)            (369)            0.486036   
26                        (26)            (370)            0.486036   
27                        (26)            (371)            0.486036   
28                        (26)            (372)            0.486036   
29                        (26)            (373)            0.486036   
...                        ...              ...                 ...   
94219840            (376, 374)  (371, 372, 375)            0.510971   
94219841            (376, 375)  (371, 372, 374)            0.448035   
94219842  (376, 371, 373, 374)            (375)            0.491123   
94219843  (376, 371, 373, 375)            (374)            0.421704   
94219844  (376, 371, 374, 375)            (373)            0.421205   
94219845  (376, 373, 374, 375)            (371)            0.430281   
94219846       (376, 371, 373)       (374, 375)            0.495312   
94219847       (376, 371, 374)       (373, 375)            0.493417   
94219848       (376, 371, 375)       (373, 374)            0.424796   
94219849       (376, 373, 374)       (371, 375)            0.501995   
94219850       (376, 373, 375)       (371, 374)            0.434470   
94219851       (376, 374, 375)       (371, 373)            0.436764   
94219852            (376, 371)  (373, 374, 375)            0.501297   
94219853            (376, 373)  (371, 374, 375)            0.510174   
94219854            (376, 374)  (371, 373, 375)            0.510971   
94219855            (376, 375)  (371, 373, 374)            0.448035   
94219856  (376, 372, 373, 374)            (375)            0.495212   
94219857  (376, 372, 373, 375)            (374)            0.425593   
94219858  (376, 372, 374, 375)            (373)            0.425893   
94219859  (376, 373, 374, 375)            (372)            0.430281   
94219860       (376, 372, 373)       (374, 375)            0.498903   
94219861       (376, 372, 374)       (373, 375)            0.498005   
94219862       (376, 372, 375)       (373, 374)            0.429384   
94219863       (376, 373, 374)       (372, 375)            0.501995   
94219864       (376, 373, 375)       (372, 

# Results

In [16]:
# Let's sort the results based on the leverage (interest)
res = rules_conf.sort_values(by=['interest'], ascending=False)

In [25]:
res = res[res['interest']>=intr]
res

antecedents      consequents  antecedent support  \
49224288  (330, 331, 326)       (328, 329)            0.421205   
28477828  (330, 283, 326)       (328, 329)            0.407241   
49224471  (330, 338, 326)       (328, 329)            0.422103   
49224705  (330, 347, 326)       (328, 329)            0.421704   
49225225  (330, 326, 367)       (328, 329)            0.419509   
49224445  (337, 330, 326)       (328, 329)            0.422302   
49224419  (336, 330, 326)       (328, 329)            0.422202   
49224497  (330, 339, 326)       (328, 329)            0.422103   
49224988  (326, 330, 358)       (328, 329)            0.421604   
49224962  (330, 357, 326)       (328, 329)            0.421404   
49225329  (330, 371, 326)       (328, 329)            0.411530   
49224575  (342, 330, 326)       (328, 329)            0.422402   
49224523  (330, 340, 326)       (328, 329)            0.422202   
49224884  (354, 330, 326)       (328, 329)            0.421704   
49225014  (330, 326, 359)       (328, 329)            0.421604   
49225251  (368, 330, 326)       (328, 329)            0.419011   
49224601  (330, 326, 343)       (328, 329)            0.422402   
49224627  (344, 330, 326)       (328, 329)            0.422202   
49224549  (330, 341, 326)       (328, 329)            0.422202   
49225043  (360, 330, 326)       (328, 329)            0.421305   
49224393  (330, 326, 335)       (328, 329)            0.422701   
49224809  (330, 326, 351)       (328, 329)            0.422103   
49225095  (330, 326, 362)       (328, 329)            0.421105   
49225147  (330, 364, 326)       (328, 329)            0.420706   
49224315  (330, 332, 326)       (328, 329)            0.422900   
49224367  (334, 330, 326)       (328, 329)            0.422801   
49224783  (350, 330, 326)       (328, 329)            0.422202   
49225069  (361, 330, 326)       (328, 329)            0.421604   
49225121  (330, 363, 326)       (328, 329)            0.421305   
49225173  (330, 365, 326)       (328, 329)            0.420606   
...                   ...              ...                 ...   
8645400        (333, 279)  (360, 281, 330)            0.436465   
28479372       (283, 326)  (352, 330, 328)            0.414622   
8644342        (332, 279)  (281, 330, 350)            0.436166   
8644276        (332, 279)  (281, 330, 347)            0.436166   
8644476        (332, 279)  (281, 330, 356)            0.436166   
8627249        (329, 279)  (281, 364, 335)            0.425194   
49267085       (331, 326)  (328, 362, 333)            0.426192   
49271540  (331, 349, 326)       (328, 339)            0.424397   
49226134  (329, 331, 326)       (328, 362)            0.421404   
49272484  (331, 340, 326)       (328, 359)            0.424197   
49234414  (329, 340, 326)       (328, 366)            0.420706   
49237370  (344, 329, 326)       (328, 366)            0.420706   
49231042  (336, 329, 326)       (328, 366)            0.420706   
49233610  (329, 339, 326)       (328, 366)            0.420706   
49277816  (331, 357, 326)       (328, 349)            0.423399   
23599644       (282, 326)  (328, 330, 347)            0.410034   
23599419       (282, 326)  (328, 330, 338)            0.410034   
51890359  (346, 332, 327)       (329, 330)            0.451227   
51890545  (353, 332, 327)       (329, 330)            0.451227   
49276921  (331, 363, 326)       (328, 347)            0.423100   
49271771  (339, 331, 326)       (328, 359)            0.424097   
49230029  (329, 326, 335)       (328, 362)            0.421305   
8648140   (330, 279, 343)       (337, 281)            0.427588   
51890339       (332, 327)  (345, 329, 330)            0.453122   
28478190       (344, 326)  (328, 329, 283)            0.430580   
49284719  (332, 326, 351)       (328, 334)            0.425494   
49242391  (329, 353, 326)       (328, 362)            0.421105   
49266121       (332, 326)  (328, 331, 358)            0.427987   
49277956  (331, 363, 326)       (328, 349)            

## Interpretation of Results
So we have 75 rules of our interest (high *support, confidence* and *interest*). Let's dig deeper in these results.

In [18]:
df_buckets.columns = df_buckets.columns.map(str)
df_buckets.columns

Index(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
       ...
       '990', '991', '992', '993', '994', '995', '996', '997', '998', '999'],
      dtype='object', length=1000)

In [29]:
specs = np.array(df_buckets[(df_buckets['331'] == True) & 
           (df_buckets['363'] == True) & 
           (df_buckets['326'] == True) & 
           (df_buckets['328'] == True) &
           (df_buckets['249'] == True)].reset_index()['index'])

seqs = list(pep_spec.keys())

In [28]:
# List of peptides having this common itemset

for kk in specs:
    print (seqs[kk])

AASWLTHQGSFHGAFR
AEQGSGPVSGEKDVVFLIDGSEGVR
AQFGQPEILLGTIPGAGGTQR
ATISNPVTGALETVHYR
CVDVRPGYCYTALANGR
DAILDALENLSGDELKK
DDLFNTNATIVATLTAACAQHCPEAMVCIIANPVNSTIPITAEVFKK
DGYVQVEEYIADLYSEEPGEEEPAWVQTER
DLAGALDEKELIEK
DQDCDKFNQCGTCTEFK
DQVDSAVQELLQLK
EATEYEIELYGISR
EFVMQEAQQNDDVSK
EGTGTETPMIGDR
EKFSPLTANLMNLLAENGR
ELVDEEADEAQELLSQAENWQR
ELVVVDTPGIFDTEVPDADTQR
EVSSGSEETFSYSDFLR
EVTLPLPRPQDVK
FKLEENYNMNDALYK
FLYFSNWLHGDIR
FNTESQGCNHEEDAGVR
GFLFGPSLAQELGVGCVLIR
GLYGINEDVFLSVPCILGQNGISDVVK
GQHPLSIQPMGVSFTAENEGTYYGPPGR
HEAGDMMGGHAIR
HSFGPLDYESLQQELALK
IAIVGAGIGGTSSAYYLR
IDGNLVIRPYTPVSSDDDKGFVDLVVK
IGCDQHTSCPVGQTCCPSLK
IGDLQSQIVSLLK
IIVYNQPYTGYEK
IKPHLMSQEVPEDWDKQPVK
ILDDVKEDDYLNFILFSTDVTTWKDHLVQATPANLK
ILIPGINDAVAPVTDEEHALYDHIDFDMEEFAK
ILVNSQYTASVFK
IQFNESFAEMNR
KAEEEELASPLR
KDPSSNDPLTETIR
KIEFEQFLPMMQAISNNKDQGGYEDFVEGLR
KNDGSFGAWLHR
LEENYNMNDALYK
LLPQVMYLDGYDR
LQAALDNEAGGRPAMEPGNGSLDLGGDAAGR
LREAEETQNSLQAECDQYR
LSAAAQSTVYAFSARPLTGGEPVSLGSLR
LSGTGEALPSVPGVVLGPEELPVEVLQFHPTVDGVLVSTAGK
LVQAQYWHDPI